larq_zoo/literature/real_to_bin_nets.py

```diff
@@ -0,0 +1,490 @@
+            3, strides=2, padding="same", name=f"{name}_pool"
+        )(x)
+
+    def last_block(self, x: tf.Tensor, name: str = "") -> tf.Tensor:
+        """Last block, shared across ResNet, StrongBaselineNet and Real-to-Bin nets."""
+
+        x = tf.keras.layers.GlobalAvgPool2D(name=f"{name}_global_pool")(x)
+        x = tf.keras.layers.Dense(
+            self.num_classes, activation=None, name=f"{name}_logits",
+        )(x)
+        return tf.keras.layers.Softmax(name=f"{name}_probs", dtype=tf.float32)(x)
+
+    def block(self, x, downsample=False, name=""):
+        """Main network block
+
+        This block differs between ResNet and StrongBaseline / Real-to-Bin Nets.
+        It is implemented by the ResNet18 and StrongBaselineNet subclasses.
+        """
+
+        raise NotImplementedError
+
+    def shortcut_connection(
+        self, x: tf.Tensor, name: str, in_channels: int, out_channels: int
+    ) -> tf.Tensor:
+        if in_channels == out_channels:
+            return x
+        x = tf.keras.layers.AvgPool2D(
+            2, strides=2, padding="same", name=f"{name}_shortcut_pool"
+        )(x)
+        x = tf.keras.layers.Conv2D(
+            out_channels,
+            kernel_size=1,
+            kernel_initializer=self.kernel_initializer,
+            kernel_regularizer=self.kernel_regularizer,
+            use_bias=False,
+            name=f"{name}_shortcut_conv2d",
+        )(x)
+        return tf.keras.layers.BatchNormalization(
+            momentum=self.momentum, name=f"{name}_shortcut_batch_norm"
+        )(x)
+
+    def build(self) -> tf.keras.models.Model:
+        """Build the model."""
+
+        x = self.first_block(
+            self.image_input,
+            name=f"{self.model_name}_block_1",
+            use_prelu=isinstance(self, StrongBaselineNetFactory),
+        )
+        for block in range(2, 10):
+            x = self.block(
+                x,
+                name=f"{self.model_name}_block_{block}",
+                downsample=block % 2 == 0 and block > 3,
+            )
+
+        if self.include_top:
+            x = self.last_block(x, name=f"{self.model_name}_block_10")
+
+        model = tf.keras.Model(
+            inputs=self.image_input, outputs=x, name=self.model_name,
+        )
+
+        # Load weights.
+        if self.weights == "imagenet":
+            model.load_weights(self._get_weights_path())
+        elif self.weights is not None:
+            model.load_weights(self.weights)
+        return model
+
+    def _get_weights_path(self):
+        raise NotImplementedError(f"No stored weights for {self.model_name}")
+
+
+@factory
+class StrongBaselineNetFactory(_SharedBaseFactory):
+    """Constructor for the strong baseline network (Section 4.1 of Martinez et al.)."""
+
+    scaling_r: int = 8
+
+    input_quantizer: QuantizerType = Field(None)
+    kernel_quantizer: QuantizerType = Field(None)
+
+    class LearnedRescaleLayer(tf.keras.layers.Layer):
```

Comment:
I think I would define this layer outside of the class, but this is just personal preference.

Reply:
I can see your point (in fact, that is how it was earlier). The reasoning for having it here is for when people have read the paper and then look at our implementation:
R2B net is architecturally just a StrongBaselineNet which has a different conv output scaling strategy. (And StrongBaselineNet is just a ResnetE18 with some modifications). The current implementation is I think not the best software design but it is the most pedagogical way of defining things to understand the paper (and if you understand the paper to understand the code).